tests/TestCompiler.hs

```diff
@@ -161,48 +108,39 @@ assertCompiles supportModules supportExterns supportForeigns inputFiles outputFi
           Just (ExitFailure _, _, err) -> return $ Just err
           Nothing -> return $ Just "Couldn't find node.js executable"
 
-assertCompilesWithWarnings
+assertGolden
   :: [P.Module]
   -> [P.ExternsFile]
   -> M.Map P.ModuleName FilePath
-  -> [FilePath]
-  -> [String]
-  -> Expectation
-assertCompilesWithWarnings supportModules supportExterns supportForeigns inputFiles shouldWarnWith =
-  assert supportModules supportExterns supportForeigns inputFiles checkMain $ \e ->
-    case e of
-      Left errs ->
-        return . Just . P.prettyPrintMultipleErrors P.defaultPPEOptions $ errs
-      Right warnings ->
-        return
-          . fmap (printAllWarnings warnings)
-          $ checkShouldFailWith shouldWarnWith warnings
-
-  where
-  printAllWarnings warnings =
-    (<> "\n\n" <> P.prettyPrintMultipleErrors P.defaultPPEOptions warnings)
-
-assertDoesNotCompile
+  -> [[FilePath]]
+  -> [TestTree]
+assertGolden supportModules supportExterns supportForeigns testCases =
+  flip fmap testCases $ \files ->
+    let mainFile = head files
+    in
+      goldenVsString
+      (takeBaseName mainFile)
+      (replaceExtension mainFile ".out")
+      (BS.fromStrict . T.encodeUtf8 . T.pack <$> printErrorOrWarning supportModules supportExterns supportForeigns files)
+
+printErrorOrWarning
```

Comment:
We shouldn't really be smushing errors and warnings together like this. With the current code, these tests are verifying that everything in the `failing` directory produces at least one error, and that everything in the `warning` directory produces only warnings; if I'm reading this correctly, we're losing that here.

Reply:
What do you mean with smushing errors and warnings together? I'll check the code again later to see if I get what you mean :)

Reply:
What I mean is that this function discards the information of whether a particular MultipleErrors was raised as an error or as a warning.

Reply:
Mm I see, I thought it wasn't that important since we were checking the error/warning output and I assumed it was different in the two cases. 
Isn't it going to be checked by the directives test once I put them back in anyways?

Reply:
Oh of course, yes it would be.

Reply:
By the way part of the reason I think it’s really important to test this properly is because the exit code depends on it too,  and most workflows (and especially CI) rely on the exit codes to be correct.